opt: make grouping columns optional in the GroupBy/DistinctOn orderings

RaduBerinde opened the pull request:
#### opt: adjust cost of sort

The cost of sort is proportional to the amount of columns we are
sorting on. A single sort on three columns has the same cost as two
sorts (one on a single column and one on two columns), even though
it's clearly preferable to sort once.

This change adjusts the formula so it has a term that accounts for the
non-comparison cost.

A follow-up commit adds testcases with DISTINCT ON which would have
two sorts without this change.

Release note: None

#### opt: make DISTINCT ON columns optional in the internal ordering

The purpose of the DistinctOn internal ordering is to determine the
intra-group ordering; this change makes all the grouping columns be
optional columns in the ordering.

Release note: None

#### opt: add grouping columns as optional columns to GroupBy ordering

Similar to DistinctOn, the internal ordering in GroupBy is used for
intra-group ordering. We add the grouping columns as optional.

Release note: None

#### opt: output internal-ordering field in expressions

Now that the internal ordering is not always identical to the required
ordering on the input, display the internal ordering for Limit,
Offset, and GroupBy variations.

Release note: None

RaduBerinde committed:
opt: make DISTINCT ON columns optional in the internal ordering

The purpose of the DistinctOn internal ordering is to determine the
intra-group ordering; this change makes all the grouping columns be
optional columns in the ordering.

Release note: None

RaduBerinde committed:
opt: adjust cost of sort

The cost of sort is proportional to the amount of columns we are
sorting on. A single sort on three columns has the same cost as two
sorts (one on a single column and one on two columns), even though
it's clearly preferable to sort once.

This change adjusts the formula so it has a term that accounts for the
non-comparison cost.

A follow-up commit adds testcases with DISTINCT ON which would have
two sorts without this change.

Release note: None

RaduBerinde committed:
opt: add grouping columns as optional columns to GroupBy ordering

Similar to DistinctOn, the internal ordering in GroupBy is used for
intra-group ordering. We add the grouping columns as optional.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/28332)
<!-- Reviewable:end -->

RaduBerinde committed:
opt: output internal-ordering field in expressions

Now that the internal ordering is not always identical to the required
ordering on the input, display the internal ordering for Limit,
Offset, and GroupBy variations.

Release note: None

RaduBerinde commented:
Fixed a bug in Intersection, and added a commit that adds back the ordering in the expression output (as `internal-ordering`, and also for limit/offset).

rytaft reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 4 of 4 files at r1, 6 of 9 files at r2, 5 of 5 files at r3, 3 of 3 files at r4, 22 of 22 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/28332#-:-LJL4JiY67rxo_WU4MxL:bnfp4nl)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

___
*[pkg/sql/logictest/testdata/logic_test/distsql_distinct_on, line 95 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/28332#-LJKy-Tm2WfoF2CUBEK4:-LJKy-Tm2WfoF2CUBEK5:b390i4s) ([raw file](https://github.com/cockroachdb/cockroach/blob/98562aaf2590e2802e8571f550abbf37a8dfe4c9/pkg/sql/logictest/testdata/logic_test/distsql_distinct_on#L95)):*
> ```
> 4 1 6
> 
> query III partialsort(1)
> ```

I think I've asked this question before.... but why is partialsort needed here?
___
*[pkg/sql/opt/props/ordering_choice_test.go, line 113 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/28332#-LJL0W-Y7QTDAvMTBCWp:-LJL0W-Y7QTDAvMTBCWq:boonl2m) ([raw file](https://github.com/cockroachdb/cockroach/blob/98562aaf2590e2802e8571f550abbf37a8dfe4c9/pkg/sql/opt/props/ordering_choice_test.go#L113)):*
> ```Go
> 		{left: "+(1|2)", right: "+1 opt(2)", expected: false},
> 		{left: "+(1|2),-(3|4)", right: "+(1|2),-(3|4),+5", expected: false},
> ```

[nit] why the extra line?
___
*[pkg/sql/opt/props/ordering_choice_test.go, line 132 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/28332#-LJL1ca0EcuO1N1MiLbW:-LJL1ca0EcuO1N1MiLbX:b10ca2m) ([raw file](https://github.com/cockroachdb/cockroach/blob/98562aaf2590e2802e8571f550abbf37a8dfe4c9/pkg/sql/opt/props/ordering_choice_test.go#L132)):*
> ```Go
> 
> // TestOrderingChoice_Intersection tests Intersects and Intersection.
> func TestOrderingChoice_Intersection(t *testing.T) {
> ```

Nice tests!
___
*[pkg/sql/opt/xform/testdata/physprops/ordering, line 938 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/28332#-LJL2P5A4LJ7Pd0-DgC5:-LJL2P5A4LJ7Pd0-DgC6:b-qwhfnu) ([raw file](https://github.com/cockroachdb/cockroach/blob/98562aaf2590e2802e8571f550abbf37a8dfe4c9/pkg/sql/opt/xform/testdata/physprops/ordering#L938)):*
> ```
> 
> # There is no ordering that satisfies both the intra-group ordering of b+ and the
> # inter-group ordering of c+; we have to sort twice.
> ```

I think this comment isn't quite right... the intra-group ordering is c+, and the inter-group ordering is a+, right?


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
TFTR!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/28332#-:-LJL66pw0UkJt0VDIxyK:b-814g8k)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

___
*[pkg/sql/logictest/testdata/logic_test/distsql_distinct_on, line 95 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/28332#-LJKy-Tm2WfoF2CUBEK4:-LJL4zwQDxHyU7oH5gJ5:br581xm) ([raw file](https://github.com/cockroachdb/cockroach/blob/98562aaf2590e2802e8571f550abbf37a8dfe4c9/pkg/sql/logictest/testdata/logic_test/distsql_distinct_on#L95)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

I think I've asked this question before.... but why is partialsort needed here?
</blockquote></details>

We're only specifying an order for `x`. So the rows `1 1 NULL`, `1 1 2` and `1 2 1` can appear in any order.
___
*[pkg/sql/opt/props/ordering_choice_test.go, line 113 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/28332#-LJL0W-Y7QTDAvMTBCWp:-LJL5XQ49E64VjHYkDac:bss4a2b) ([raw file](https://github.com/cockroachdb/cockroach/blob/98562aaf2590e2802e8571f550abbf37a8dfe4c9/pkg/sql/opt/props/ordering_choice_test.go#L113)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

[nit] why the extra line?
</blockquote></details>

Removed.
___
*[pkg/sql/opt/xform/testdata/physprops/ordering, line 938 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/28332#-LJL2P5A4LJ7Pd0-DgC5:-LJL5j-f112fksze6jzj:bygn2t8) ([raw file](https://github.com/cockroachdb/cockroach/blob/98562aaf2590e2802e8571f550abbf37a8dfe4c9/pkg/sql/opt/xform/testdata/physprops/ordering#L938)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

I think this comment isn't quite right... the intra-group ordering is c+, and the inter-group ordering is a+, right?
</blockquote></details>

Good catch, I switched around the columns in the query and didn't update the comment.


<!-- Sent from Reviewable.io -->

rytaft reviewed (approved):
> Reviewed 1 of 25 files at r6, 2 of 3 files at r7, 22 of 22 files at r8.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/28332)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

___
*[pkg/sql/logictest/testdata/logic_test/distsql_distinct_on, line 95 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/28332#-LJKy-Tm2WfoF2CUBEK4:-LJL6cEV5h7kCOxTpS3X:b4c9lyr) ([raw file](https://github.com/cockroachdb/cockroach/blob/98562aaf2590e2802e8571f550abbf37a8dfe4c9/pkg/sql/logictest/testdata/logic_test/distsql_distinct_on#L95)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

We're only specifying an order for `x`. So the rows `1 1 NULL`, `1 1 2` and `1 2 1` can appear in any order.
</blockquote></details>

:thumbsup: 


<!-- Sent from Reviewable.io -->

andy-kimball reviewed (approved):
> Reviewed 4 of 4 files at r1, 9 of 9 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/28332)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

___
*[pkg/sql/opt/props/ordering_choice.go, line 210 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/28332#-LJK6LNO94cYjDfyXBKw:-LJK6LNO94cYjDfyXBKx:bbnrz10) ([raw file](https://github.com/cockroachdb/cockroach/blob/e3dafecd99d184c006947248c815b879b56b7999/pkg/sql/opt/props/ordering_choice.go#L210)):*
> ```Go
> 		if !oc.Optional.Contains(int(ord[i].ID())) {
> 			oc.Columns = append(oc.Columns, OrderingColumnChoice{
> 				Group:      util.MakeFastIntSet(int(ord[i].ID())),
> ```

Seems unnecessarily expensive to call `MakeFastIntSet` for a single value, since it will trigger creation of garbage slice. Can we add constructor that takes single column (I've wanted that for a long time)? Alternatively, a call to `Add` like we do in `FromOrdering` would avoid an unnecessary allocation.


<!-- Sent from Reviewable.io -->

andy-kimball reviewed (commented):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/28332#-:-LJLXKG_5l6I6RIWp7J4:bg7igkv)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale)



<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/28332#-:-LJM-vxe17-dLdYsF1TL:b-ydi9ay)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale)

___
*[pkg/sql/opt/props/ordering_choice.go, line 210 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/28332#-LJK6LNO94cYjDfyXBKw:-LJLyDe_2UoVRIKYqJh6:b-y0ngjw) ([raw file](https://github.com/cockroachdb/cockroach/blob/e3dafecd99d184c006947248c815b879b56b7999/pkg/sql/opt/props/ordering_choice.go#L210)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

Seems unnecessarily expensive to call `MakeFastIntSet` for a single value, since it will trigger creation of garbage slice. Can we add constructor that takes single column (I've wanted that for a long time)? Alternatively, a call to `Add` like we do in `FromOrdering` would avoid an unnecessary allocation.
</blockquote></details>

Discussed offline, the slice doesn't get allocated (it's set up on the stack).


<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Flake (#28352).

bors r+

RaduBerinde commented:
Merge skew, thanks bors!

bors+

RaduBerinde commented:
![image](https://user-images.githubusercontent.com/16544120/43843408-7f0a76f8-9af5-11e8-97cd-b3f82dd9eb32.png)

rytaft commented:
:laughing: :laughing: :laughing: 





<!-- Sent from Reviewable.io -->
